### PR TITLE
refactor(dialog): new LuDialogService approach for dialog boxes

### PR DESCRIPTION
**Pull request by Supamiu**
## Description

New approach for dialog boxes inside Lucca Front 🎉 

This first iteration adds the `LuDialogService` class that you can provide using the `provideLuDialog()` function in order to open dialog boxes.

🚧 # TODO

- Documentation !

-----

In order to open a dialog box using a component, you can use:

## Service-based approach:

```ts
#dialog = inject(LuDialogService);
...
openDialog(): void {
    const dialogRef = this.#dialog.open({
        // content can also be a TemplateRef
        content: MyDialogComponent
    });
    // If you want to get the result from the dialog
    dialogRef.result$.subscribe(...)
    // If you want to know when the dialog has been dismissed (closed using the `dismiss()` method)
    dialogRef.dismissed$.subscribe(...)
    // If you want to know when the dialog has been closed, no matter if it dismissed or not
    dialogRef.closed$.subscribe(...)
}
```

Inside the dialog box component, there's helper structure components to help you create good dialogs effortlessly, here is an example component:

```ts
@Component({
	selector: 'test-dialog',
	template: `
                <!-- If you have a form, just replace this div with a form and keep the class on it so the form becomes parent of the whole dialog box -->
		<div class="dialog-inside">
		        <!-- lu-dialog-header will take care of generating an id and giving it to the dialog box as ariaLabelledby attribute, it'll also add a close button and hook it on the canClose config -->
			<lu-dialog-header>I'm a test dialog ! Hello {{ world }}</lu-dialog-header>
			
			<!-- content ensures you're using the right class for the content container -->
			<lu-dialog-content>
				I'm the content of the dialog !
				You can put anything you want in here
			</lu-dialog-content>
			
			<!-- footer will put everything in a visually separated place, with all the footer tooling that you need in order to have spacing between buttons, additional info, etc -->
			<lu-dialog-footer>
				<div class="footer-content">Additional footer text</div>
				<div class="footer-actions">
					<button type="button" luButton (click)="close()">Confirm</button>
					<button type="button" luButton="text" (click)="close()">Cancel</button>
				</div>
			</lu-dialog-footer>
		</div>
	`,
	imports: [
		ButtonComponent,
		DialogHeaderComponent,
		DialogFooterComponent,
		DialogContentComponent,
	],
	standalone: true,
})
class TestDialogContent {
	world = injectDialogData<string>();
	ref = injectDialogRef<number>();

	close(): void {
	        // We have to provide a number to the close method here because we injected the reference as number above.
		this.ref.close(0);
	}
}
```

## Template-driven approach

```html
<button luButton [luDialogOpen]="dialogTpl">Open Template-driven Dialog</button>

<ng-template #dialogTpl>
	<lu-dialog #dialog>
		<lu-dialog-header>I'm a template driven header</lu-dialog-header>

		<lu-dialog-content>
			I'm the content of the dialog !
		</lu-dialog-content>

		<lu-dialog-footer>
			<div class="footer-content">Footer text !</div>
			<div class="footer-actions">
				<button type="button" luButton luDialogClose>Confirm using Directive</button>
				<button type="button" luButton luDialogDismiss>Dismiss using Directive</button>
				<button type="button" luButton (click)="dialog.close()">Close from dialog (also has dismiss)</button>
				<button type="button" luButton (click)="doSomethingWithRef(dialog.dialogRef)">Do something with ref</button>
			</div>
		</lu-dialog-footer>
	</lu-dialog>
</ng-template>
```

ℹ️ These structure components are heavily recommended but you can create a dialog box without them, be aware that it'll hurt accessibility if you don't provide an `ariaLabel` as replacement for the header, you'll also have to create the close "X" and place it yourself.

Documentation is being written in parallel which is why I won't put every detail here but this basic explanation should be enough to play around with it 🙂. Additionaly, here's the config interface:

```ts
interface BaseLuDialogConfig<C> {
	/**
	 * The component or tamplate to put iside the dialog container
	 */
	content: ComponentType<C> | TemplateRef<C>;

	/**
	 * Data to pass to the component, will be required if the component used `injectDialogData` in a field, specifying the data that's needed.
	 */
	data: LuDialogData<C>;

	/**
	 * Should we put a backdrop? Defaults to true
	 */
	backdrop?: boolean;

	/**
	 * Can this dialog box be dismissed by clicking on the backdrop or pressing escape?
	 *
	 * Defaults to true, setting this to false will also remove the close button in the header
	 * if you're using `lu-dialog-header`.
	 */
	dismissible?: boolean;

	/**
	 * Can be used if you don't have a header or aren't using the default one, to set an aria-label
	 * instead of aria-labelledby on the dialog container
	 */
	ariaLabel?: string;

	/**
	 * This provides a way to override ano of the configuration parameters expected by cdk's `Dialog.open` method.
	 *
	 * WARNING: this will take full authority and override with your values no matter what's computed by LuDialogService,
	 * use with caution.
	 */
	cdkConfigOverride?: DialogConfig<C>;

	/**
	 * A hook function to determine if the current dialog can be closed or not, if provided.
	 *
	 * @param comp the instance of the component that's inside the dialog box.
	 */
	canClose?: (comp: C) => boolean | Observable<boolean>;
}
```
-----

### CODE REVIEW

**Comment by LuccaIntegration** [comment]
:woman_cook: https://lucca-front.lucca.tech/PR-2557/storybook

**Comment by CCNET-iLucca** [comment]
[Tests d'interfaces](https://qa.lucca.tech/12994/ui-diff)

**Comment by CCNET-iLucca** [comment]
[Tests d'interfaces](https://qa.lucca.tech/12998/ui-diff)

File: packages/ng/dialog/dialog.service.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+					}
+					if (config.canClose) {
+						const canClose = config.canClose(cdkRef.componentInstance);
+						if (isObservable(canClose)) {
+							return canClose.pipe(first());
+						}
+						return of(canClose);
+					}
+					return of(true);
+				}),
```
Comment (lamstutz) [inline packages/ng/dialog/dialog.service.ts:51]: je sais que c'est pas terminé mais au cas ou ...
il faudrais ajouter un truc du genre takeUntil(ref.close$) (ou via la subsription) ici non ?

**Comment by LuccaIntegration** [comment]
:woman_cook: https://lucca-front.lucca.tech/PR-2557/storybook

**Comment by CCNET-iLucca** [comment]
[Tests d'interfaces](https://qa.lucca.tech/13027/ui-diff)

**Comment by LuccaIntegration** [comment]
:woman_cook: https://lucca-front.lucca.tech/PR-2557/storybook

**Comment by CCNET-iLucca** [comment]
[Tests d'interfaces](https://qa.lucca.tech/13034/ui-diff)

File: packages/ng/dialog/model/dialog-config.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+	[K in keyof T]: T[K] extends { [dialogData]: infer D } ? D : never;
+}[keyof T];
+
+export type LuDialogResult<C> = {
+	[K in keyof C]: C[K] extends { [dialogResult]: infer R } ? R : never;
+}[keyof C];
+
+interface BaseLuDialogConfig<C> {
+	/**
+	 * The component or tamplate to put iside the dialog container
```
Comment (Ptit-Biscuit) [inline packages/ng/dialog/model/dialog-config.ts:19]: Petites typo sur `template` et `inside` :)

File: packages/ng/dialog/model/dialog-config.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+	dismissible?: boolean;
+
+	/**
+	 * Can be used if you don't have a header or aren't using the default one, to set an aria-label
+	 * instead of aria-labelledby on the dialog container
+	 */
+	ariaLabel?: string;
+
+	/**
+	 * This provides a way to override ano of the configuration parameters expected by cdk's `Dialog.open` method.
```
Comment (Ptit-Biscuit) [inline packages/ng/dialog/model/dialog-config.ts:48]: `any` ?

**Comment by LuccaIntegration** [comment]
:woman_cook: https://lucca-front.lucca.tech/PR-2557/storybook

**Comment by CCNET-iLucca** [comment]
[Tests d'interfaces](https://qa.lucca.tech/13050/ui-diff)

File: packages/ng/dialog/dialog/dialog.component.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
+import { ChangeDetectionStrategy, Component, ElementRef, inject, OnInit, ViewEncapsulation } from '@angular/core';
+import { LuDialogRef } from '../model';
+
+@Component({
+	selector: 'lu-dialog',
+	standalone: true,
+	imports: [],
+	template: '<ng-content></ng-content>',
```
Comment (lamstutz) [inline packages/ng/dialog/dialog/dialog.component.ts:8]: ici on se permet de mettre du inlineTemplate, on pourrais faire la meme chose sur le dialogFooter component ? ou c'est peut-etre pour etre iso avec le header ?

Comment (Supamiu) [reply packages/ng/dialog/dialog/dialog.component.ts:8]: Tu as raison, j'ai fait le changement 👍

**Comment by LuccaIntegration** [comment]
:woman_cook: https://lucca-front.lucca.tech/PR-2557/storybook

**Comment by CCNET-iLucca** [comment]
[Tests d'interfaces](https://qa.lucca.tech/13131/ui-diff)

**Comment by LuccaIntegration** [comment]
:woman_cook: https://lucca-front.lucca.tech/PR-2557/storybook

File: packages/ng/dialog/dialog-content/dialog-content.component.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
+import { ChangeDetectionStrategy, Component, ViewEncapsulation } from '@angular/core';
+import { CheckboxInputComponent } from '@lucca-front/ng/forms';
+import { FormFieldComponent } from '@lucca-front/ng/form-field';
+
+@Component({
+	selector: 'lu-dialog-content',
+	standalone: true,
+	imports: [CheckboxInputComponent, FormFieldComponent],
+	templateUrl: './dialog-content.component.html',
+	styleUrl: './dialog-content.component.scss',
```
Comment (lamstutz) [inline packages/ng/dialog/dialog-content/dialog-content.component.ts:10]:
```suggestion
	template: '<ng-content></ng-content>',
```

**Comment by CCNET-iLucca** [comment]
[Tests d'interfaces](https://qa.lucca.tech/13176/ui-diff)

**Review by jbosi — DISMISSED** [review]

File: packages/ng/dialog/dialog.service.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+				takeUntil(luDialogRef.closed$),
+			)
+			.subscribe((canClose) => {
+				if (canClose) {
+					cdkRef.close(DISMISSED_VALUE);
+				}
+			});
+		return luDialogRef;
+	}
+}
```
Comment (lamstutz) [inline packages/ng/dialog/dialog.service.ts:59]: je trouverais ca plus clair
```suggestion
if (config.dismissible === false) {
		this.handleDismiss(cdkRef)
			.pipe(
				switchMap(() => {
					const canClose = config.canClose?.(cdkRef.componentInstance) ?? true;
					const canClose$ = isObservable(canClose) ? canClose : of(canClose);
					return canClose$.pipe(take(1));
				}),
				filter(Boolean)
				takeUntil(luDialogRef.closed$),
			)
			.subscribe(() => {
				cdkRef.close(DISMISSED_VALUE);
			});
			}
		return luDialogRef;
	}
	
	#handleDismiss(cdkRef){
	 return merge(cdkRef.backdropClick, cdkRef.keydownEvents.pipe(filter((e) => e.key === 'Escape' && !e.defaultPrevented)));
	}
	
}
```

**Review by lamstutz — CHANGES_REQUESTED** [review]

**Comment by LuccaIntegration** [comment]
:woman_cook: https://lucca-front.lucca.tech/PR-2557/storybook

**Comment by CCNET-iLucca** [comment]
[Tests d'interfaces](https://qa.lucca.tech/13196/ui-diff)

File: packages/ng/dialog/dialog.service.ts (hunk 6 above)
Comment (Supamiu) [reply packages/ng/dialog/dialog.service.ts:59]: J'ai du mal à comprendre l'utilité de mettre ça dans une méthode en dessous, virer la logique entièrement si `dismissible === false` fait sens (j'imagine que c'est une partie de ce que tu voulais dire ici) mais cette méthode me semble non nécessaire.

**Comment by LuccaIntegration** [comment]
:woman_cook: https://lucca-front.lucca.tech/PR-2557/storybook

**Comment by CCNET-iLucca** [comment]
[Tests d'interfaces](https://qa.lucca.tech/13207/ui-diff)

File: packages/ng/dialog/dialog-content/dialog-content.component.ts [hunk 7]
```diff
@@ -0,0 +1,14 @@
+import { ChangeDetectionStrategy, Component, ViewEncapsulation } from '@angular/core';
+
+@Component({
+	selector: 'lu-dialog-content',
+	standalone: true,
+	imports: [],
```
Comment (shajz) [inline packages/ng/dialog/dialog-content/dialog-content.component.ts:6]: no need? dito for all other occurrences

File: packages/ng/dialog/dialog-header/dialog-header.component.ts [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+	dismissible = this.#ref.config.dismissible !== false;
+
+	close(): void {
+		this.#ref.dismiss();
+	}
+
+	ngOnInit(): void {
+		// Using setTimeout here to make sure this will be handled in the next Cd cycle, not the current one.
+		setTimeout(() => {
```
Comment (shajz) [inline packages/ng/dialog/dialog-header/dialog-header.component.ts:41]: Is Angular missing a lifecycle hook for this use case ? 🤯

Comment (Supamiu) [reply packages/ng/dialog/dialog-header/dialog-header.component.ts:41]: It's more that if we change anything at this exact time, we'll get an error due to expression changed while it was checked, which could result in unapplied changes in prod environment.

**Comment by LuccaIntegration** [comment]
:woman_cook: https://lucca-front.lucca.tech/PR-2557/storybook

**Comment by CCNET-iLucca** [comment]
[Tests d'interfaces](https://qa.lucca.tech/13216/ui-diff)

**Review by shajz — DISMISSED** [review]

**Comment by LuccaIntegration** [comment]
:woman_cook: https://lucca-front.lucca.tech/PR-2557/storybook

**Comment by CCNET-iLucca** [comment]
[Tests d'interfaces](https://qa.lucca.tech/13242/ui-diff)

**Comment by LuccaIntegration** [comment]
:woman_cook: https://lucca-front.lucca.tech/PR-2557/storybook

File: packages/ng/dialog/dialog/dialog.component.ts [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+	close(): void {
+		this.dialogRef.close();
+	}
+
+	dismiss(): void {
+		this.dialogRef.dismiss();
+	}
+
+	ngOnInit(): void {
+		if (!this.dialogRef.config.cdkConfigOverride?.autoFocus) {
```
Comment (vvalentin-lucca) [inline packages/ng/dialog/dialog/dialog.component.ts:29]: By default, the focus should be on the close button. You want to be able to override it, but it must remain an exception that you activate deliberately.

**Comment by CCNET-iLucca** [comment]
[Tests d'interfaces](https://qa.lucca.tech/13282/ui-diff)

File: packages/ng/dialog/dialog.service.ts (hunk 6 above)
Comment (Supamiu) [reply packages/ng/dialog/dialog.service.ts]: Je comprend mais ici on écoute les évènements, la méthode devrait plutôt s'appeler `listenToBackdropandEscape` car aucune logique liée au dismiss n'est faite ici.

**Comment by LuccaIntegration** [comment]
:woman_cook: https://lucca-front.lucca.tech/PR-2557/storybook

**Comment by CCNET-iLucca** [comment]
[Tests d'interfaces](https://qa.lucca.tech/13300/ui-diff)

**Comment by LuccaIntegration** [comment]
:woman_cook: https://lucca-front.lucca.tech/PR-2557/storybook

**Comment by CCNET-iLucca** [comment]
[Tests d'interfaces](https://qa.lucca.tech/13303/ui-diff)

**Review by jeremie-lucca — APPROVED** [review]

**Review by Ptit-Biscuit — APPROVED** [review]

**Pull request merged**